trio/_tools/gen_exports.py

```diff
@@ -105,42 +102,78 @@ def create_passthrough_args(funcdef: ast.FunctionDef | ast.AsyncFunctionDef) ->
-    """
-    # imported to check that `subprocess` calls to black will succeed
+    :raises SystemExit: If black failed."""
+    # imported to check that `subprocess` calls will succeed
     import black  # noqa: F401
 
     # Black has an undocumented API, but it doesn't easily allow reading configuration from
     # pyproject.toml, and simultaneously pass in / receive the code as a string.
     # https://github.com/psf/black/issues/779
-    try:
-        result = subprocess.run(
-            # "-" as a filename = use stdin, return on stdout.
-            [sys.executable, "-m", "black", "--stdin-filename", file.path, "-"],
-            input=source,
-            capture_output=True,
-            encoding="utf8",
-            check=True,
-        )
-    except subprocess.CalledProcessError as exc:
-        print("Failed to run black!")
-        traceback.print_exception(type(exc), exc, exc.__traceback__)
+    result = subprocess.run(
+        # "-" as a filename = use stdin, return on stdout.
+        [sys.executable, "-m", "black", "--stdin-filename", file.path, "-"],
+        input=source,
+        capture_output=True,
+        encoding="utf8",
+    )
+
+    if result.returncode != 0:
+        return False, f"Failed to run black!\n{result.stderr}"
+    return True, result.stdout
+
+
+def run_ruff(file: File, source: str) -> tuple[bool, str]:
+    """Run ruff on the specified file, returning tuple of success and result string.
+
+    :raises ImportError: If ruff is not installed
+    :raises SystemExit: If ruff failed."""
+    # imported to check that `subprocess` calls will succeed
+    import ruff  # noqa: F401
+
+    result = subprocess.run(
+        # "-" as a filename = use stdin, return on stdout.
+        [
+            sys.executable,
+            "-m",
+            "ruff",
+            "--fix-only",
+            "--format=text",
+            "--stdin-filename",
+            file.path,
+            "-",
+        ],
+        input=source,
+        capture_output=True,
+        encoding="utf8",
+    )
+
+    if result.returncode != 0 or result.stderr:
+        return False, f"Failed to run ruff!\n{result.stderr}"
+    return True, result.stdout
+
+
+def run_linters(file: File, source: str) -> str:
+    """Run isort and black on the specified file, returning the new source.
+
+    :raises ImportError: If either is not installed
+    :raises SystemExit: If either failed.
+    """
+
+    success, response = run_black(file, source)
+    if not success:
+        print(response)
         sys.exit(1)
-    # isort does have a public API, makes things easy.
-    try:
-        isort_res = isort.api.sort_code_string(
-            result.stdout,
-            file_path=file.path,
-        )
-    except isort.exceptions.ISortError as exc:
-        print("Failed to run isort!")
-        traceback.print_exception(type(exc), exc, exc.__traceback__)
+
+    success, response = run_ruff(file, response)
+    if not success:
+        print(response)
```

Comment:
`# pragma: no cover`, unless you want to write a test that triggers it

Reply:
Added in d364693